src/plugins/workspace/public/components/workspace_creator/types.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { App, WorkspacePermissionMode } from '../../../../../core/public';
+
+export interface WorkspaceFeature extends Pick<App, 'dependencies'> {
+  id: string;
+  name: string;
+}
+
+export interface WorkspaceFeatureGroup {
+  name: string;
+  features: WorkspaceFeature[];
+}
+export interface WorkspaceFormSubmitData {
+  name: string;
+  description?: string;
+  features?: string[];
+  color?: string;
+  icon?: string;
+  defaultVISTheme?: string;
+  permissions: WorkspacePermissionSetting[];
+}
+
+export interface WorkspaceFormData extends WorkspaceFormSubmitData {
+  id: string;
+  reserved?: boolean;
+}
+
+export type WorkspaceFormErrors = Omit<
+  { [key in keyof WorkspaceFormData]?: string },
+  'permissions'
+> & {
+  userPermissions?: string[];
+  groupPermissions?: string[];
+};
+
+export enum WorkspacePermissionItemType {
+  User = 'user',
+  Group = 'group',
+}
+
+export interface UserPermissionSetting {
+  type: WorkspacePermissionItemType.User;
+  userId: string;
+  modes: WorkspacePermissionMode[];
+}
+
+export interface GroupPermissionSetting {
+  type: WorkspacePermissionItemType.Group;
+  group: string;
+  modes: WorkspacePermissionMode[];
+}
+
+export type WorkspacePermissionSetting = UserPermissionSetting | GroupPermissionSetting;
+
+// when editing, attributes could be undefined in workspace form
+export type WorkspaceFormEditingData = Partial<
+  Omit<WorkspaceFormSubmitData, 'permissions'> & {
+    userPermissions: Array<Partial<UserPermissionSetting>>;
+    groupPermissions: Array<Partial<GroupPermissionSetting>>;
+  }
+>;
+
+export type UserOrGroupPermissionEditingData = Array<
```

Comment:
I'd rename it to `PermissionFieldData`, that becomes

```typescript
export type PermissionFieldData = {
  id?: string; 
  modes?: WorkspacePermissionMode[]
}
```

And in `getUserAndGroupPermissions` function, I will do
```typescript
  const userPermissions: PermissionFieldData[] = [];
  const groupPermissions: PermissionFieldData[] = [];
```

Reply:
How about declaring both variables in `PermissionFieldData` and create a new type for editing permission data:

`type PermissionEditingData = Array<Partial<PermissionFieldData>>; `